### DISCUSSION TITLE
Issue Decoding trace data

### DISCUSSION THREAD

**Discussion by vishalkadam18** [apache/skywalking#12512] [Q&A] [answered]
Dear Team,

We are using Apache SkyWalking-apm tool with Elasticsearch and cannot decode the trace data . Base64 decoding doesn’t reveal a recognizable pattern or structure. We need guidance on what encoding is used, how to decode and structure the data, and how to identify and separate different parts of the data. 

Are there specific tools or methods recommended for this task? 

Thanks!
Vishal

**Comment by wankai123** [comment, marked as answer]
Hi, you need to follow the below code to decode the trace data:
base64 decode the data from ES and use SegmentObject to parse the result.
I think you missed `SegmentObject.parseFrom(segment.getDataBinary())`

- https://github.com/apache/skywalking/blob/9b17ff1efeab7a20c870839f59eb0e6af485cd3f/oap-server/server-storage-plugin/storage-elasticsearch-plugin/src/main/java/org/apache/skywalking/oap/server/storage/plugin/elasticsearch/base/ElasticSearchConverter.java#L53
- https://github.com/apache/skywalking/blob/9b17ff1efeab7a20c870839f59eb0e6af485cd3f/oap-server/server-core/src/main/java/org/apache/skywalking/oap/server/core/query/TraceQueryService.java#L172
